src/Workspaces/Core/Portable/Indentation/AbstractIndentationService.AbstractIndenter.cs

```diff
@@ -108,8 +115,36 @@ public IndentationResult GetDesiredIndentation(FormattingOptions.IndentStyle ind
                     token, previousNonWhitespaceOrPreprocessorLine, lastNonWhitespacePosition);
             }
 
-            protected abstract IndentationResult GetDesiredIndentationWorker(
-                SyntaxToken token, TextLine previousLine, int lastNonWhitespacePosition);
+            public bool TryGetSmartTokenIndentation(out IndentationResult indentationResult)
+            {
+                indentationResult = default;
+                if (!ShouldUseTokenIndenter(out var token))
+                {
+                    return false;
+                }
+
+                // var root = document.GetSyntaxRootSynchronously(cancellationToken);
+                var sourceText = Tree.GetText(CancellationToken);
+
+                var formatter = CreateSmartTokenFormatter();
+                var changes = formatter.FormatTokenAsync(Document.Project.Solution.Workspace, token, CancellationToken)
+                                       .WaitAndGetResult(CancellationToken);
+
+                var updatedSourceText = sourceText.WithChanges(changes);
+                if (LineToBeIndented.LineNumber < updatedSourceText.Lines.Count)
```

Comment:
I believe we do this kind of thing in test to see whether things worked correctly, but smart token formatter is formatter not indenter meaning it is supposed to change buffer...

Reply:
None of this actually ends up changing the buffer.  That's for the clients of these types to do.  All that's happening here though is using the results of these types (immutably) to try to determine the result to pass back.